### ISSUE TITLE
Allow user-defined monads for defining `Spec`s

### ISSUE THREAD

**Issue by srid** [hspec/hspec#676]
contd. from https://github.com/hspec/hspec/pull/667#issuecomment-1109869075

Instead of `SpecWith a`, if we had a `SpecWith m a` - it would allow us to use custom monads for, say, managing project-specific test configuration.

In the normal case, one would use `SpecWith IO a`, but for advanced cases I might want to create a test configuration like the below:

```hs
-- Specify where to create golden files for golden tests
data TestConf = TestConf { tcGoldenBasePath :: FilePath } 
```

And then modify it for certain sub-trees. Assuming `m` is `StateT TestConf IO`, it would look like:

```haskell
oldConfig <- get
modify $ \c -> c { tcGoldenBasePath = "./foo" }
describe "foo" $ ...
put oldConfig
describe "bar" $ ...
```

Here, "foo" sub-tree will use the new config, but all other specs would use the initial config.

**Comment by srid** [comment]
cc @TotallyNotChase ... hopefully I've described above the general feature request behind our use of GoldenConf sufficiently.

**Comment by TotallyNotChase** [comment]
yes, I'd really appreciate this feature. Some way of "introspection" is really the feature I particularly wanted, but this should be more general and allow even cooler usecases.

**Comment by sol** [comment]
I think this describes a specific implementation rather than a user story.

From what I understand, you want to store/retrieve configuration values in the spec monad. Is that correct? If that’s the case, then I think there are simply implementations to achieve that, e.g. something similar to what the `vault` package provides.

One potential downside of parametrizing `SpecM` Is that it hurts compatibility.

**Comment by srid** [comment]
> you want to store/retrieve configuration values in the spec monad. Is that correct?

Ya. I'd say store, retrieve **and** modify on a sub-tree (ie., `describe`, but not `it`) basis. @TotallyNotChase Let me know if I missed anything.

**Comment by sol** [comment]
I think in the simplest case we can just store a function in our monad stack that essentially maps type representations to Maybe values of some existential. On set/modify we the simply modify that function by wrapping it into a new function. That will be less efficient than using a proper `Map`, but likely good enough for our use cases.

From a users perspective that will give us:

```
setValue :: Typeable c => v -> SpecM c ()
getValue :: Typeable c => SpecM a (Maybe c)
```

A user can then use these to implement custom primitives for his config type, e.g.

```
data MyConfi = ...

defaultConfig :: MyConfig
defaultConfig = ...

setConfig :: MyConfig -> SpecM a ()
setConfig = setValue

getConfig :: SpecM a MyConfig
getConfig = fromMaybe defaultConfig <$> getValue

modifyConfig f = getConfig >>= setConfig . f
```

@srid Will this work for your use case?

**Comment by parsonsmatt** [comment]
This feels pretty doable with the hooks API - `before` is essentially `local` for a given part of the spec tree.

**Comment by sol** [comment]
@parsonsmatt Hspec makes a distinction between “constructing a spec tree” and “executing a spec tree”.  So the question is “when exactly do you want to access additional configuration?”.

- At construction time we can’t access hooks. We can run IO actions during construction time though. More on this later.
- At execution time hooks are executed as the test runner traverses the tree of spec items. One thing to note here is that not necessarily all spec items from the constructed spec tree will be executed (think focused items, filtering, etc.).

from what I understand, this issue is about configuration at construction time. @srid @TotallyNotChase please correct me if I’m wrong.

I haven’t had time to work on this yet. However, we have `runIO`. This allows you to work around the current limitation if you really want to. There are several options:

- use the `unsafePerformIO` hack
- Store/access your configuration in/from an environment variable
- Store/access your configuration in/from a temporary file

@srid @TotallyNotChase somewhat unrelated, another thing that you can do with `runIO` is e.g. create a spec item for each file in a directory, or even a spec tree based on a filesystem (sub)tree. Just mentioning this, as it may be relevant for golden tests.
